### ISSUE TITLE
Possible performance issue in points cache

### ISSUE THREAD

**Issue by skulonen** [apluslms/a-plus#882]
(From #880, delayed feedback)

https://github.com/apluslms/a-plus/blob/8a852f917e815876892ffa82eeca3e08d82d4882/exercise/cache/points.py#L178-L181

This could potentially be a performance bottleneck. Querying all exercises one-by-one from the database is typically slow. Though, we can't be certain without performance testing.

If it were faster to fetch all exercises in one query from the database, then this line could pick the right exercise from a dictionary with the id, or something like that. The dictionary would be filled once with all exercises in one query.

**Comment by skulonen** [comment]
Actually, the performance of DEADLINE_ALL can be greatly improved by rewriting the `get_latest_deadline` to not fetch the `exercise` object before fetching the deadline, and to cache the result between the calls to `RevealRule.is_revealed` and `RevealRule.get_reveal_time`. With this improvement, it's not strictly necessary to figure out the caching for the maximum deadline deviation, although it could still be done.

https://github.com/apluslms/a-plus/blob/4312c99fbb257d190c37b4589bd28671e32c97cf/exercise/reveal_states.py#L85

**Comment by skulonen** [comment]
I did some more deviation testing. I set all exercises in the A+ Manual course to use the DEADLINE_ALL reveal mode, and added 500 dummy users to the course, and gave a deadline deviation for every user in every exercise.

- The total time taken by finding one user's personal deadline and personal max submissions for all exercises now takes **~80** ms (before adding deviations: **~10** ms).
- The total time taken by evaluating the DEADLINE_ALL reveal rule for all exercises now takes **~80** ms (before adding deviations: **~60** ms).
- The total time taken by CachedPoints._generate_data now takes **~300** ms (before adding deviations: **~220** ms).

There are a couple of additional optimizations that could be done:
1. Cache deviations separately from CachedPoints. Since deviations are updated much more rarely than CachedPoints is regenerated, that could improve the performance of CachedPoints.
2. Implement some sort of batch evaluation of reveal rules, that could take advantage of the ability to query for the maximum deadline deviations for all exercises using one query.

Or maybe this is "fast enough" already?
